### PR TITLE
Optimize FCL move shape

### PR DESCRIPTION
**Pull request by captain-yoshi**
### Description
Related to #3599.

Updating a MoveIt Collision object shape poses takes 0.3 seconds for an object with 3 meshes with a total of 48,064 triangles. 

When the FCL collision environment is notified by a World::CREATE | MOVE_SHAPE | ADD_SHAPE | REMOVE_SHAPE, the object is reconstructed from scratch as seen [here](https://github.com/moveit/moveit/blob/master/moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp#L424-L443).

This PR only updates the FCL object transforms when encountering the `MOVE_SHAPE` action. The time is reduced to an average of 16 microseconds.

### CODE REVIEW

**Review by simonschmeisser — COMMENTED** [review]
Nice!

I think inverting the logic of your error handling would make it easier to read.

Also a test would be great to move some object in or out of collision and verify it.

**Comment by captain-yoshi** [comment]
I will invert the logic and try to add tests. Do you know if the failing test is failing due to these changes ?

```bash
[  FAILED  ] FCLCollisionCheckPanda/CollisionDetectorPandaTest/0.RobotWorldCollision_1, where TypeParam = collision_detection::CollisionDetectorAllocatorFCL (13 ms)
```

**Comment by captain-yoshi** [comment]
@simonschmeisser Ok, this was not intuitive ! We got to recompute the AABB and unregister/register the FCL object from the broad phase manager so that the collision detector works.

Added 2 more tests where the object is not in collision.

| Height Z axis (m) | In Collision ? |
|-------------------|----------------|
| 0.3               | True           |
| 0.55              | False          |
| 0.6               | True           |
| 0.85              | False          |
<p align="middle">
  
  
  
  
</p>

**Comment by captain-yoshi** [comment]
It now takes an average of 1.5 milliseconds to update the 3 meshes.

File: moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp [hunk 1]
```diff
@@ -434,6 +434,36 @@ void CollisionEnvFCL::notifyObjectChange(const ObjectConstPtr& obj, World::Actio
...
+      ROS_ERROR_NAMED(LOGNAME, "Cannot move shapes of unknown FCL object: '%s'", obj->id_.c_str());
+      return;
+    }
+
+    if (obj->global_shape_poses_.size() != it->second.collision_objects_.size())
+    {
+      ROS_ERROR_NAMED(LOGNAME,
+                      "Cannot move shapes, shape size mismatch between FCL object and world object: '%s'. Respectively "
+                      "%zu and %zu.",
+                      obj->id_.c_str(), it->second.collision_objects_.size(), it->second.collision_objects_.size());
```
Comment (simonschmeisser) [inline moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp:451]:
```suggestion
                      obj->id_.c_str(), it->second.collision_objects_.size(), it->second.collision_objects_.size());
                      return;
```

File: moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp [hunk 2]
```diff
@@ -434,6 +434,36 @@ void CollisionEnvFCL::notifyObjectChange(const ObjectConstPtr& obj, World::Actio
...
+    for (std::size_t i = 0; i < it->second.collision_objects_.size(); ++i)
+    {
+      it->second.collision_objects_[i]->setTransform(transform2fcl(obj->global_shape_poses_[i]));
+
+      // compute AABB, order matters
+      it->second.collision_geometry_[i]->collision_geometry_->computeLocalAABB();
+      it->second.collision_objects_[i]->computeAABB();
+    }
+
+    // Reregister the FCL object into the broad phase collision manager
```
Comment (simonschmeisser) [inline moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp:463]: Ideally the comment would explain why this needs to happen. Is there no API for moving objects in fcl? (I know it's not implemented for the manager actually used)

**Review by simonschmeisser — CHANGES_REQUESTED** [review]

File: moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp (hunk 1 above)
Comment (captain-yoshi) [reply moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp:451]: Oups thanks!

**Review by simonschmeisser — APPROVED** [review]

File: moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp (hunk 2 above)
Comment (captain-yoshi) [reply moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp:464]: I will try to see why this is needed (Actually I just repeated what was done [here](https://github.com/moveit/moveit/blob/master/moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp#L366C23-L399)).

I can't find any API to move the object in FCL, only the `CollisionEnvFCL::updateFCLObject` which reconstruct's the object...

There is 3 options to update the broadphase collision manager:
```
    for (std::size_t i = 0; i < it->second.collision_objects_.size(); ++i)
    {
      it->second.collision_objects_[i]->setTransform(transform2fcl(obj->global_shape_poses_[i]));

      // compute AABB, order matters
      it->second.collision_geometry_[i]->collision_geometry_->computeLocalAABB();
      it->second.collision_objects_[i]->computeAABB();

      // Option #1 (18 usec)
      // manager_->update(it->second.collision_objects_[i].get());
    }
    // Option #2 (15 usec)
    // manager_->update();
    
    // Option #3 (7 usec)
    // it->second.unregisterFrom(manager_.get());
    // it->second.registerTo(manager_.get());
```
The fastest way seems to be the original fix I proposed.

**Comment by captain-yoshi** [comment]
Did some benchmarks:

| Update Manager methods            | Move Primitive Shape (usec) \* | Move 3 Mesh (usec) \*\*|
|-----------------------------------|------------------------|---------------|
| reregister                        | 7                      | 2583          |
| manager update                    | 15                     | 2894          |
| manager update(collision_objects) | 18                     | 2613          |

\*&nbsp;*Primitive Shape: World::moveObject*
\*\**PlanningScene::processCollisionObjectMove*

This is not representative of a real world example, because the scene would have many more collision objects (meshes, primitives or planes). With 3 meshes we see that actually the `manager update(collision_objects)` is actually faster then `manager update`.

By checking the the [FCL Broadphase Dynamic AABB](https://github.com/flexible-collision-library/fcl/blob/master/include/fcl/broadphase/broadphase_dynamic_AABB_tree-inl.h) we can see the 3 methods.

### Method 1: registerObject/unregiterObject
The `table` variable is an `std::unordered_map` which has an average O(1) lookup (depending on the hash function) and an average of O(M/N) to iterate over keys when the key is not found. The `dtree` variable is a hierarchy tree structure from [fcl](https://github.com/flexible-collision-library/fcl/blob/master/include/fcl/broadphase/detail/hierarchy_tree.h).

I would argue that these method should be used when we only **move** shapes, because it will be inserted at the same place. Thus the dtree does not need to be rebalanced (See Method#2) ? 
```cpp
void DynamicAABBTreeCollisionManager<S>::registerObject(CollisionObject<S>* obj)
{
  DynamicAABBNode* node = dtree.insert(obj->getAABB(), obj);
  table[obj] = node;
}

template <typename S>
void DynamicAABBTreeCollisionManager<S>::unregisterObject(CollisionObject<S>* obj)
{
  DynamicAABBNode* node = table[obj];
  table.erase(obj);
  dtree.remove(node);
}
```

### Method 2: update
We iterate over every elements of the table to get the AABB. After this, the dtree is rebalanced.

You can see [here](https://github.com/moveit/moveit/blob/master/moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp#L163-L164) and [here](https://github.com/moveit/moveit/blob/master/moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp#L398) that this method was maybe used but has been commented in the code in favor of the register method.

Another guess for it not being used is that the dtree update changes the structure of the tree even if the bounding volume has not been changed https://github.com/flexible-collision-library/fcl/issues/368. It is called by `balanceIncremental`. 

```cpp
template <typename S>
FCL_EXPORT
void DynamicAABBTreeCollisionManager<S>::setup()
{
  if(!setup_)
  {
    int num = dtree.size();
    if(num == 0)
    {
      setup_ = true;
      return;
    }

    int height = dtree.getMaxHeight();


    if(height - std::log((S)num) / std::log(2.0) < max_tree_nonbalanced_level)
      dtree.balanceIncremental(tree_incremental_balance_pass);
    else
      dtree.balanceTopdown();

    setup_ = true;
  }
}

template <typename S>
void DynamicAABBTreeCollisionManager<S>::update()
{
  for(auto it = table.cbegin(); it != table.cend(); ++it)
  {
    CollisionObject<S>* obj = it->first;
    DynamicAABBNode* node = it->second;
    node->bv = obj->getAABB();
  }

  dtree.refit();
  setup_ = false;

  setup();
}
```

### Method 3: update specific objects
We iterate over only the specified collision objects of the table. After this, the dtree is rebalanced.

I would recommend this method when adding a **new** collision object because the dtree should be rebalanced at some point ? 

```cpp
void DynamicAABBTreeCollisionManager<S>::update_(CollisionObject<S>* updated_obj)
{
  const auto it = table.find(updated_obj);
  if(it != table.end())
  {
    DynamicAABBNode* node = it->second;
    if(!node->bv.equal(updated_obj->getAABB()))
      dtree.update(node, updated_obj->getAABB());
  }
  setup_ = false;
}

template <typename S>
void DynamicAABBTreeCollisionManager<S>::update(const std::vector<CollisionObject<S>*>& updated_objs)
{
  for(size_t i = 0, size = updated_objs.size(); i < size; ++i)
    update_(updated_objs[i]);
  setup();
}
```
----
### TL;DR
I'm not sure if we need to rebalance the dtree at some point. I would think that this is needed when **adding** a new collision object or after every N objects. We should use the `DynamicAABBTreeCollisionManager<S>::update(const std::vector<CollisionObject<S>*>& updated_objs)` method so that the `dtree.balanceIncremental` is only trigered once.

When **moving** a collision object, I think that the unregister/register pattern should be used. When inserting a leaf node in the dtree, the bounding volumes are updated as necessary. This seems to corroborate the benchmark results.

My bandwidth for this has exploded and cannot explore things more. I will add a better comment for why the register is needed.

**Comment by captain-yoshi** [comment]
Actually there are 4 ways to update the broadphase manager (I updated the previous comments with a fourth option). The new update method is this one: `update(const std::vector<CollisionObject<S>*>& updated_objs)`.

So updating it this way is actually faster then the un/register method. So I will update my fix with this method instead.

image
\**Every scene had 100+ meshes in total*

**Comment by captain-yoshi** [comment]
Ok... so actually using `manager_->update(fcl_raw_objs)` does not update the AABB as it fails the test... Maybe there is an error with FCL ? The `manager_->update()` works. While I'm at it I will investigate just to make sure.

Will surely revert to the original fix (reregister) because it is slightly faster then the `manager_->update()` and this is what is done when creating an object.


image

**Comment by captain-yoshi** [comment]
Found my bug, was wrongfully initialising the vector. Also I was probably not using enough samples. Bumped to 3000. Based on the benchmarks below, I don't think that a method is really superior then the other.

So we probably should stick with the reregistering method for updating the AABB. This is what is actually used when creating an object. By using this method, the tree is not rebalanced, but it's not clear what is the effect based on the benchmarks. Furthermore, the `update()` method has been commented out, so it might be for a reason.

I have re-pushed the initial fix with better comments including the git commit.

### Benchmark 1: Moving a primitive shape 3000 times by 0.25 meters
This test depends on the load on my PC, and was pretty inconsistent between runs. It's really fast so that maybe why: Q1 and Q3 around 370 nanoseconds. It benchmarks the `World::moveObject` method.


### Benchmark 2: Moving meshes 20@25k + 1@161k
The planning scene has an additional 96 meshes @25k that are not moved. It benchmarks the `PlanningScene::processCollisionObjectMove` method. I don't know if we can really choose a method based on the graphs...

**Comment by captain-yoshi** [comment]
@simonschmeisser Can you retrigger the CI, it fails to fetch dependencies.

**Review by rhaschke — APPROVED** [review]

**Comment by simonschmeisser** [comment]
Unfortunately #3694 still crashes for me. Given that the cache is `thread_local`, moving would need to happen in every thread or the id/hash would need to change or something to invalidate the cache. I'll think further about it and revert this in our branch for the next release of our software
